File: project/CCNet.VBDemos.Plugin/SourceControl/IndexFileHistoryParser.vb
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Xml.Linq
Imports ThoughtWorks.CruiseControl.Core
Imports ThoughtWorks.CruiseControl.Core.Sourcecontrol

Public Class IndexFileHistoryParser
    Implements IHistoryParser


    Public Function Parse(ByVal history As TextReader, _
                          ByVal from As Date, ByVal [to] As Date) As Modification() _
                          Implements IHistoryParser.Parse
        Dim modifications = New List(Of Modification)()
        Dim document = XDocument.Load(history)
        Dim fromDate = from.ToString("s")
        Dim toDate = [to].ToString("s")
        For Each change As XElement In document.Descendants("change")
            Dim changeDate = DateTime.Parse(change.Attribute("date").Value)
            If (changeDate >= from) And (changeDate <= [to]) Then
                Dim newModification = New Modification()
                newModification.FileName = change.Attribute("file").Value
                newModification.ModifiedTime = changeDate
                newModification.FolderName = change.Attribute("folder").Value
                newModification.Type = change.Attribute("type").Value
                modifications.Add(newModification)
            End If
        Next
        Return modifications.ToArray()
    End Function
End Class

File: project/CCNet.CSharpDemos.Plugin/SourceControl/IndexFileHistoryParser.cs
namespace CCNet.CSharpDemos.Plugin.SourceControl
{
    using System;
    using System.Collections.Generic;
    using System.IO;
    using System.Xml.Linq;
    using ThoughtWorks.CruiseControl.Core;
    using ThoughtWorks.CruiseControl.Core.Sourcecontrol;

    public class IndexFileHistoryParser
        : IHistoryParser
    {
        public Modification[] Parse(TextReader history, DateTime from, DateTime to)
        {
            var modifications = new List<Modification>();
            var document = XDocument.Load(history);
            var fromDate = from.ToString("s");
            var toDate = to.ToString("s");
            foreach (XElement change in document.Descendants("change"))
            {
                var date = DateTime.Parse(
                    change.Attribute("date").Value);
                if ((date >= from) &&
                    (date <= to))
                {
                    var modification = new Modification
                    {
                        FileName = change.Attribute("file").Value,
                        ModifiedTime = date,
                        FolderName = change.Attribute("folder").Value,
                        Type = change.Attribute("type").Value
                    };
                    modifications.Add(modification);
                }
            }

            return modifications.ToArray();
        }
    }
}
